engine/src/main/java/io/zeebe/engine/nwe/event/IntermediateCatchEventProcessor.java

```diff
@@ -0,0 +1,104 @@
+
+public class IntermediateCatchEventProcessor
+    implements BpmnElementProcessor<ExecutableCatchEventElement> {
+
+  private final BpmnEventSubscriptionBehavior eventSubscriptionBehavior;
+  private final BpmnIncidentBehavior incidentBehavior;
+  private final BpmnStateBehavior stateBehavior;
+  private final BpmnStateTransitionBehavior stateTransitionBehavior;
+  private final BpmnVariableMappingBehavior variableMappingBehavior;
+
+  public IntermediateCatchEventProcessor(final BpmnBehaviors bpmnBehaviors) {
+    eventSubscriptionBehavior = bpmnBehaviors.eventSubscriptionBehavior();
+    incidentBehavior = bpmnBehaviors.incidentBehavior();
+    stateBehavior = bpmnBehaviors.stateBehavior();
+    stateTransitionBehavior = bpmnBehaviors.stateTransitionBehavior();
+    variableMappingBehavior = bpmnBehaviors.variableMappingBehavior();
+  }
+
+  @Override
+  public Class<ExecutableCatchEventElement> getType() {
+    return ExecutableCatchEventElement.class;
+  }
+
+  @Override
+  public void onActivating(
+      final ExecutableCatchEventElement element, final BpmnElementContext context) {
+    variableMappingBehavior
+        .applyInputMappings(context, element)
+        .flatMap(ok -> eventSubscriptionBehavior.subscribeToEvents(element, context))
+        .ifRightOrLeft(
+            ok -> stateTransitionBehavior.transitionToActivated(context),
+            failure -> incidentBehavior.createIncident(failure, context));
+  }
+
+  @Override
+  public void onActivated(
+      final ExecutableCatchEventElement element, final BpmnElementContext context) {
+    if (element.isNone()) {
+      stateTransitionBehavior.transitionToCompleting(context);
+    }
+  }
+
+  @Override
+  public void onCompleting(
+      final ExecutableCatchEventElement element, final BpmnElementContext context) {
+    variableMappingBehavior
+        .applyOutputMappings(context, element)
+        .ifRightOrLeft(
+            ok -> {
+              eventSubscriptionBehavior.unsubscribeFromEvents(context);
+              stateTransitionBehavior.transitionToCompleted(context);
+            },
+            failure -> incidentBehavior.createIncident(failure, context));
+  }
+
+  @Override
+  public void onCompleted(
+      final ExecutableCatchEventElement element, final BpmnElementContext context) {
+    stateTransitionBehavior.takeOutgoingSequenceFlows(element, context);
+    stateBehavior.consumeToken(context);
+    stateBehavior.removeElementInstance(context);
+  }
+
+  @Override
+  public void onTerminating(
+      final ExecutableCatchEventElement element, final BpmnElementContext context) {
+    eventSubscriptionBehavior.unsubscribeFromEvents(context);
+    stateTransitionBehavior.transitionToTerminated(context);
+  }
+
+  @Override
+  public void onTerminated(
+      final ExecutableCatchEventElement element, final BpmnElementContext context) {
+    eventSubscriptionBehavior.publishTriggeredBoundaryEvent(context);
```

Comment:
Nope. An intermediate catch event may not have a boundary event attached to it :grin: